### PR TITLE
Issue #2747 - Using linalg everywhere

### PR DESCRIPTION
**Pull request by ckousik**
This PR is a work in progress which will transition shogun to using linalg for linear algebra operations. Minimal unit tests will be written for required classes.
#2747

### CODE REVIEW

**Comment by lisitsyn** [comment]
Thanks for the patch!

Please take a look at the Travis CI status, the tests are apparently failing.

@karlnapf can you take a look?

**Comment by ckousik** [comment]
@lisitsyn I have been through the travis build. The failure is caused by the style checker. I'll push another commit shortly to fix it. Sorry :(

File: tests/unit/distribution/Gaussian_unittest.cc [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+TEST(Gaussian, log_pdf_single_1d)
+{
+	SGVector<float64_t> mean(1), point(1);
+	SGMatrix<float64_t> cov(1,1);
+	
+	mean[0] = 0;
+	point[0] = 0;
+	cov(0,0) = 1.0;
+
+	CGaussian *gauss = new CGaussian(mean, cov, FULL);
```
Comment (karlnapf) [inline tests/unit/distribution/Gaussian_unittest.cc]: you could do `auto gaussian = some<CGaussian>(mean, cov, FULL)`

File: tests/unit/distribution/Gaussian_unittest.cc [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+	SG_UNREF(gauss);
+}
+
+TEST(Gaussian, log_pdf_multiple_independent_2d)
+{
+	SGVector<float64_t> mean(2);
+	SGMatrix<float64_t> cov(2,2);
+
+	mean[0] = mean[1] = 0;
+	// Random variables are independent
```
Comment (karlnapf) [inline tests/unit/distribution/Gaussian_unittest.cc]: no need for this comment

File: tests/unit/distribution/Gaussian_unittest.cc [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+	cov(0,0) = 1.0;
+	cov(0,1) = 0.0;
+	cov(1,0) = 0.0;
+	cov(1,1) = 1.0;
+
+	auto gauss = new CGaussian(mean, cov, FULL);
+	// Find the log of the distribution function at the mean
```
Comment (karlnapf) [inline tests/unit/distribution/Gaussian_unittest.cc]: no need for comment

File: tests/unit/distribution/Gaussian_unittest.cc [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+	// position.
+	float64_t log_pdf = gauss->compute_log_PDF(mean);
+	EXPECT_NEAR(log_pdf, -1.83787706641, 1e-8);
```
Comment (karlnapf) [inline tests/unit/distribution/Gaussian_unittest.cc]: float64 is 16 digits, not just 8

File: tests/unit/distribution/Gaussian_unittest.cc [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+	SGMatrix<float64_t> data(1, 500);	
+	for(int32_t i = 0; i < 500; i++)
+		data(0, i) = CMath::randn_double();
+	auto train_features = new CDenseFeatures<float64_t>(data);
+	CGaussian *gauss = new CGaussian();
+	gauss->train(train_features);
+
+	auto mean = gauss->get_mean();
+	auto cov = gauss->get_cov();
+	EXPECT_NEAR(mean[0], 0.0, 1e-1);
```
Comment (karlnapf) [inline tests/unit/distribution/Gaussian_unittest.cc]: i would rather compute the mean by hand and check whether it is the same here.

File: tests/unit/distribution/Gaussian_unittest.cc [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+			auto rd = CMath::randn_double();
+			auto sigma = CMath::sqrt(sample_cov(i,i));
+			data(i, j) = sigma*rd + sample_mean[i];
+		}
+	}
+
+	CGaussian *gauss = new CGaussian();
+	auto train_features = new CDenseFeatures<float64_t>(data);
+	gauss->train(train_features);
+
```
Comment (karlnapf) [inline tests/unit/distribution/Gaussian_unittest.cc]: see comments above, also hold here

File: tests/unit/distribution/Gaussian_unittest.cc [hunk 7]
```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright (c) The Shogun Machine Learning Toolbox
+ * Written (w) 2014 Parijat Mazumdar
```
Comment (karlnapf) [inline tests/unit/distribution/Gaussian_unittest.cc]: you can put your name here :)

**Comment by ckousik** [comment]
@karlnapf I am getting precision errors when testing with eps set to 1e-16. I checked the DotFeatures unit tests and they are only checked for 1e-8.

**Comment by ckousik** [comment]
@karlnapf I'm also concerned about the usefulness of these tests. The Gaussian mean and covairance matrices are generated from DotFeatures, and so this feels like a more elaborate test for DotFeatures than for Gaussian.

File: tests/unit/distribution/Gaussian_unittest.cc [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+	cov(1,1) = 2.0;
+	auto gauss = some<CGaussian>(mean, cov, FULL);
+	float64_t log_pdf = gauss->compute_log_PDF(mean);
+	EXPECT_NEAR(log_pdf, -2.8108321409370021, 1e-16);
+}
+
+TEST(Gaussian, train_univariate)
+{
+	float64_t eps = 1e-8;
+	sg_rand->set_seed(1);
```
Comment (karlnapf) [inline tests/unit/distribution/Gaussian_unittest.cc]: why do you need a fixed seed when you test for the estimation? Should be ok everytime no?

File: tests/unit/distribution/Gaussian_unittest.cc [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+		data(0, i) = CMath::randn_double();
+
+	auto gauss = some<CGaussian>();
+	auto train_features = some<CDenseFeatures<float64_t>>(data);
+	gauss->train(train_features);
+
+	auto mean = gauss->get_mean();
+	auto cov = gauss->get_cov();
+
+	for (int32_t i = 0; i < sample_size; i++)
```
Comment (karlnapf) [inline tests/unit/distribution/Gaussian_unittest.cc]: for `auto i : range(sample_size)` ... minor

File: tests/unit/distribution/Gaussian_unittest.cc [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+
+TEST(Gaussian, train_univariate)
+{
+	float64_t eps = 1e-8;
+	sg_rand->set_seed(1);
+
+	SGMatrix<float64_t> data(1, 500);
+
+	int64_t sample_size = 500;
+	float64_t mn = 0.0, cv = 0.0;
```
Comment (karlnapf) [inline tests/unit/distribution/Gaussian_unittest.cc]: I dont like this variable names. What about something meaningful? mean and covariance?

Comment (karlnapf) [reply tests/unit/distribution/Gaussian_unittest.cc]: mu and Sigma?

Comment (karlnapf) [reply tests/unit/distribution/Gaussian_unittest.cc]: or "mean_est" or so

File: tests/unit/distribution/Gaussian_unittest.cc [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+	sample_mean[0] = 0.0;
+	sample_mean[1] = 2.0;
+
+	sample_cov(0,0) = 2.0;
+	sample_cov(1,1) = 4.0;
+	sample_cov(0,1) = sample_cov(1,0) = 0.0;
+
+	SGMatrix<float64_t> data(2, sample_size);
+
+	for(int32_t i = 0; i < 2; i++){
```
Comment (karlnapf) [inline tests/unit/distribution/Gaussian_unittest.cc]: @vigsterkr don't we have ways to fill arrays with random gaussian numbers?

File: tests/unit/distribution/Gaussian_unittest.cc [hunk 12]
```diff
@@ -0,0 +1,187 @@
...
+
+	auto gauss = some<CGaussian>();
+	auto train_features = some<CDenseFeatures<float64_t>>(data);
+	gauss->train(train_features);
+
+	linalg::zero(sample_mean);
+	linalg::zero(sample_cov);
+
+	for (int32_t i = 0; i < sample_size; i++)
+		train_features->add_to_dense_vec(1.0, i, sample_mean.vector, 2);
```
Comment (karlnapf) [inline tests/unit/distribution/Gaussian_unittest.cc]: you are right, this should not be in here
I think there are linalg methods

**Comment by karlnapf** [comment]
But this is already improving, let's iterate a few more times, and this will be good to merge.
THANKS! :)

**Pull request closed without merging**